tools/specialcasings.py

```diff
@@ -0,0 +1,77 @@
+
+from __future__ import print_function
+import json
+import os
+import shutil
+import urllib2
+from argparse import ArgumentParser
+
+def to_unicode_escape(codepoints):
+    return "".join(map(lambda codepoint: "\\u" + codepoint, codepoints.split(" ")))
+
+def parse(casings):
+    casings_list = casings.split("\n")
+    casings_list = filter(lambda line: len(line) > 0, casings_list)
+    casings_list = filter(lambda line: line[0] != "#", casings_list)
+    casings_list = map(lambda casing: map(lambda part: part.strip(), casing.split(";")), casings_list)
+
+    # filter out casings with conditions
+    casings_list = filter(lambda casing: len(casing) == 5, casings_list)
+
+    casings_dict_list = map(lambda casing: { "code": casing[0], "lower": casing[1], "upper": casing[3], "name": casing[4] }, casings_list)
+    for casing in casings_dict_list:
+        casing["code"] = to_unicode_escape(casing["code"])
+        casing["lower"] = to_unicode_escape(casing["lower"])
+        casing["upper"] = to_unicode_escape(casing["upper"])
+        casing["name"] = casing["name"][2:]
+    return casings_dict_list
+
+def main():
+    argparser = ArgumentParser(description = "Downloads CLDR's SpecialCasings.txt and formats it for use in Chakra")
+    argparser.add_argument(
+        "-u", "--url",
+        help = "Specify a URL to download SpecialCasings.txt from",
+        default = "ftp://ftp.unicode.org/Public/10.0.0/ucd/SpecialCasing.txt"
+    )
+    argparser.add_argument(
+        "--js",
+        help = "Exports the SpecialCasings.txt file as a Javascript file with `var SpecialCasings = (json)`",
+        nargs = "?",
+    )
+    argparser.add_argument(
+        "--json",
+        help = "Exports the SpecialCasings.txt file as a JSON file. The JSON contains an array of objects, where each object has keys for code, lower, upper, and comment",
+        nargs = "?"
+    )
+    args = argparser.parse_args()
+
+    specialcasings = urllib2.urlopen(args.url).read()
+    casings_dict_list = parse(specialcasings)
+
+    if args.js is not None:
+        with open(args.js, "w") as output:
+            output.write("""//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+// This file was generated by tools/{0} using
+// SpecialCasing.txt from {1}
+
+""".format(os.path.basename(__file__), args.url))
+            output.write("")
+            output.write("var SpecialCasings = " + json.dumps(casings_dict_list, indent = 4, sort_keys = True).replace("\\\\", "\\") + ";")
+    elif args.json is not None:
+        with open(args.json, "w") as output:
+            output.write(json.dumps(casings_dict_list, indent = 4, sort_keys = True))
```

Comment:
Why does this one not need removal of double-escapes?

Reply:
Right, but why not do that on the second output type option?

Reply:
Because I thought JSON strings didn't actually have or care about the "\u" escape sequence, so the string itself should be "\\u". Perhaps that is wrong?

Reply:
JSON definitely supports escaped characters, or it would never have gotten so ubiquitous (we'd all be using some JSON-plus-encoding format instead). As you mentioned elsewhere, it all really depends on the use case, and if nobody is depending on this JSON file for anything then it doesn't really matter.

---
In reply to: [183243067](https://github.com/Microsoft/ChakraCore/pull/5026#discussion_r183243067) [](ancestors = 183243067)